### PR TITLE
AST transforms for comprehensions

### PR DESCRIPTION
**Pull request by wcjohnson**
This PR includes a complete version of the implicit-returns PR. I will rebase after that one gets merged.

### CODE REVIEW

**Comment by wcjohnson** [comment]
@rattrayalex 

Big fixes. Closures work now!

File: src/index.js [hunk 1]
```diff
@@ -183,20 +183,56 @@ export default function (babel) {
...
+    path.resync();
 
-  function transformTerminalExpressionsIntoArrPush(path, arrId) {
-    path.resync(); // not sure if this is necessary... c/p from addImplicitReturns
+    let tailPaths = getTailExpressions(path.get("body"), allowLoops);
+    for (let tailPath of tailPaths) {
+      if (tailPath.isExpressionStatement()) {
+        visitor(tailPath);
+      } else if (tailPath.isFunctionDeclaration()) {
+        // Promote function declarations to exprs.
```
Comment (rattrayalex) [inline src/index.js]: Bah, so, here's an edge case that this might break: 

```
outer(x) ->
  now x = inner(x)

  inner() ->
    x
```

obviously, this is silly code, but a function that returns a function – and uses that function internally – needs the returned function to be declared, so that it is hoisted. Converting it to an expression breaks that. 

Okay to stick with the `return fnName` semantic from the other PR?

Comment (rattrayalex) [reply src/index.js]: If so, that should obviate this function I think...

File: src/index.js [hunk 2]
```diff
@@ -183,20 +183,56 @@ export default function (babel) {
...
+        tailPath.replaceWith(newNode);
+        visitor(tailPath);
+      }
+    }
+  }
 
-    let completionRecords = path.get("body").getCompletionRecords();
-    for (let targetPath of completionRecords) {
-      if (!targetPath.isExpressionStatement()) continue;
+  function validateComprehensionLoopBody(loopBodyPath) {
```
Comment (rattrayalex) [inline src/index.js]: nice, I like this approach

File: src/index.js [hunk 3]
```diff
@@ -183,20 +183,56 @@ export default function (babel) {
...
+    });
+  }
 
-      let arrPush = t.callExpression(
-        t.memberExpression(arrId, t.identifier("push")),
-        [targetPath.node.expression]
-      );
-      targetPath.replaceWith(arrPush);
-    }
+  function wrapComprehensionInIFE(bodyVarId, bodyVarInitializer, loopBody) {
```
Comment (rattrayalex) [inline src/index.js]: nice!

Comment (rattrayalex) [reply src/index.js]: maybe `wrapComprehensionInIife` ?

File: src/index.js [hunk 4]
```diff
@@ -632,6 +668,16 @@ export default function (babel) {
     },
   });
 
+  definePluginType("ObjectComprehension", {
+    visitor: ["loop"],
+    aliases: ["Expression"],
```
Comment (rattrayalex) [inline src/index.js]: Should probably also alias "ObjectExpression", no?

File: src/index.js [hunk 5]
```diff
@@ -883,34 +929,49 @@ export default function (babel) {
...
-              "instead, await the Comprehension (eg; `y <- [for x of xs: x]`)."
-            );
-          },
-          YieldExpression(yieldPath) {
-            throw yieldPath.buildCodeFrameError("`yield` is not allowed within Comprehensions.");
-          },
-          ReturnStatement(returnPath) {
-            throw returnPath.buildCodeFrameError("`return` is not allowed within Comprehensions.");
-          },
+        validateComprehensionLoopBody(path.get("loop.body"));
```
Comment (rattrayalex) [inline src/index.js]: I was going to say the validation could go into an `enter` while the tail expression processing goes into an `exit`, but I actually think the validation would be better off in an `exit` too. That way it'll validate `<-` properly in addition to `await` etc – thoughts?

File: src/index.js [hunk 6]
```diff
@@ -883,34 +929,49 @@ export default function (babel) {
...
+
+        const id = path.scope.generateUidIdentifier("arr");
+        visitTailExpressions(path.get("loop"), true, function(tailPath) {
```
Comment (rattrayalex) [inline src/index.js]: As noted elsewhere, I don't think `visitTailExpressions` is needed anymore – a plain `for-of` here would probably be simpler

File: src/index.js [hunk 7]
```diff
@@ -883,34 +929,49 @@ export default function (babel) {
...
+          let arrPush = t.callExpression(
+            t.memberExpression(id, t.identifier("push")),
+            [tailPath.node.expression]
+          );
+          tailPath.replaceWith(arrPush);
         });
 
-        const arrId = path.scope.generateUidIdentifier("arr");
+        path.replaceWith(wrapComprehensionInIFE(id, t.arrayExpression(), path.node.loop));
```
Comment (rattrayalex) [inline src/index.js]: nice, this should make it easier to "unroll" in the future

File: src/index.js [hunk 8]
```diff
@@ -883,34 +929,49 @@ export default function (babel) {
...
 
-        const iife = t.callExpression(fn, []);
-        path.replaceWith(iife);
+          // Only SeqExprs of length 2 are valid.
+          if (
+            (seqExpr.type !== "SequenceExpression") ||
+            (seqExpr.expressions.length !== 2)
+          ) {
+            throw tailPath.buildCodeFrameError("Object comprehensions must return" +
+            " sequence expressions of the form `key`, `value`.");
```
Comment (rattrayalex) [inline src/index.js]: How about
`Object comprehensions must return a (key, value) pair.`

**Review by rattrayalex — COMMENTED** [review]
Looking great – I think basically one real change

File: src/index.js (hunk 1 above)
Comment (wcjohnson) [reply src/index.js]: Are we sure this is broken? Look at the output code. Remember that ultimately it's the javascript engine that runs it, so as long as the output code is correct, it will work.

I think a `FunctionExpression` with a name gets pretty printed by Babel into a declaration in the final code. That means that a modern JS engine will let you call it at a higher point in the scope because that's what ES5 says. So I actually think the output is correct.

One way or another, we definitely can't just blindly insert `return` as we do with internal functions -- because remember, these exprs are going to be `replacedWith` things like `array.push` or `obj[k]=v`. I'm not sure how those two things would square with each other. We would probably have to wrap in an IFE.

My inclination is that if this is generating correct code at the engine level we shouldn't change it.

File: src/index.js (hunk 4 above)
Comment (wcjohnson) [reply src/index.js]: Yes, will change.

File: src/index.js (hunk 3 above)
Comment (wcjohnson) [reply src/index.js]: Sure.

File: src/index.js (hunk 8 above)
Comment (wcjohnson) [reply src/index.js]: Been thinking about this a bit, don't think we should say `return` because then people might be tempted to actually write `return`. Also don't think we can say "in tail position" because most people don't know what that means. Do agree that it should change a bit but not sure on final wording.

File: src/index.js (hunk 6 above)
Comment (wcjohnson) [reply src/index.js]: I wouldn't be too sure about this, see my comments above.

File: src/index.js (hunk 1 above)
Comment (wcjohnson) [reply src/index.js]: The core reason this code exists is because you can't insert anything that's not an expr into the `CallExpresion` for array.push or the `AssignExpression` for obj[k]=v without failing Babel validation. So whatever we return from here it has to be an expression, and there isn't anyway around that that I can see.

File: src/index.js (hunk 5 above)
Comment (wcjohnson) [reply src/index.js]: Not sure this will work. Regardless of enter or exit, the visitor is being called at a point-in-time -- exactly when we call `validateComprehensionLoopBody`, so if the processing hasn't been done by then it won't matter.

A possible change here is to move the whole comprehension visitor to exit.

File: src/index.js (hunk 1 above)
Comment (rattrayalex) [reply src/index.js]: Try these in your console: 

```
function outer(x) {
  x = inner();
  function inner() {
    return x + 1
  }
  return inner
}
outer(1)()
// 3
```
and
```
function outer2(x) {
  x = inner();
  return function inner() {
    return x + 1
  }
}
outer2(1)()
// Uncaught ReferenceError: inner is not defined
```

Comment (rattrayalex) [reply src/index.js]: we should be generating the former.

Comment (rattrayalex) [reply src/index.js]: > I think a FunctionExpression with a name gets pretty printed by Babel into a declaration in the final code.

I don't believe so; Babel will either wrap it in parentheses if it's not already part of some expression.

Comment (rattrayalex) [reply src/index.js]: For the expression case, we can do the same thing that we're doing in `addImplicitReturns`, which is to put the name of the function in the arr.push / obj[k] =

so for example: 
```
{ for elem x in arr: x, f() -> x }
```
becomes
```
const _obj = {};
for (...) {
  const x = arr[_i];
  function f() { return x }
  _obj[x] = f;
}
```
thoughts?

Comment (wcjohnson) [reply src/index.js]: Right, this is exactly why that code is needed!

Check out the output of the two test fixtures I just pushed. AFAICT his PR is generating the proper final js output for interior declared functions as of right now.

**Comment by wcjohnson** [comment]
@rattrayalex 

Addressed some comments, added new fixtures related to the interior declared functions. I believe the current code generation is both syntactically and semantically correct.

File: src/index.js (hunk 1 above)
Comment (wcjohnson) [reply src/index.js]: @rattrayalex Babel's output plugin already does this automatically I think... Look at the new fixtures, they correctly move the declaration up

Comment (rattrayalex) [reply src/index.js]: Right, sorry, the comparable test case for comprehensions would be: 

```
[for elem x in arr:
  y = f(x)
  f(z) -> y + z
]
```
is that right?

Comment (rattrayalex) [reply src/index.js]: maybe I'm missing something... could help if you can show a test case that fails with the FunctionExpression transform removed...

File: src/index.js (hunk 5 above)
Comment (rattrayalex) [reply src/index.js]: yeah, moving the comprehension visitors to exit I think would be the way to go

File: src/index.js (hunk 8 above)
Comment (rattrayalex) [reply src/index.js]: Hmm, good points, how about... 
`Object comprehensions must end with a (key, value) pair.`

File: src/index.js (hunk 1 above)
Comment (wcjohnson) [reply src/index.js]: @rattrayalex Nevermind, you're right, the double nested function was making it work. It doesn't work at one level of nesting.

Comment (rattrayalex) [reply src/index.js]: Cool! That makes sense

**Comment by rattrayalex** [comment]
@wcjohnson once those fixes go in, I think this will be ready to go – you might want to squash! hoping to merge shortly 😄

File: src/index.js (hunk 8 above)
Comment (wcjohnson) [reply src/index.js]: yeah that is probably the best possible, will change.

File: src/index.js (hunk 5 above)
Comment (wcjohnson) [reply src/index.js]: Moving to exit visitor caused some pretty horrific regressions. Stuff like this:
```
Error: /Users/wcj/dev/node_modules/babel-plugin-lightscript/test/fixtures/comprehensions/closure-nested-semantic.js: We don't know what to do with this node type. We were previously a Statement but we can't fit in here?
      at NodePath.insertBefore (node_modules/babel-traverse/lib/path/modification.js:61:13)
```

Comment (wcjohnson) [reply src/index.js]: @rattrayalex 

Yup, I think this can't be an exit visitor. (At least not without further extensive modifications.) The `ForInArrayStaetement` gets transformed first and that breaks the assumptions we're making about ast pretty badly.

I have a working implemenation as an enter visitor which I think I will stick with for this PR.

**Comment by wcjohnson** [comment]
@rattrayalex 

Fixes are in.

**Comment by wcjohnson** [comment]
Strike that, still an issue.

**Comment by wcjohnson** [comment]
This is a nontrivial problem for object comprehensions. The tail expression is a `SequenceExpression`, but if the second expression in the sequence is a `FunctionExpression.` It needs to be turned back into a Declaration and hoisted using the old approach I was using. Won't be done tonight.

**Comment by wcjohnson** [comment]
Or maybe it will?

File: src/index.js (hunk 5 above)
Comment (rattrayalex) [reply src/index.js]: Ah, okay, that makes sense... too bad

**Comment by rattrayalex** [comment]
I think if it starts as a `FunctionExpression` it should stay as one. The edge case I outlined for implicit returns / array comprehensions isn't one that I mean to encourage; just something that shouldn't break due to unnecessary AST mangling. 

In general, let's try to keep modification to a minimum.

**Comment by wcjohnson** [comment]
@rattrayalex 

OK, so to be clear, if someone writes:
```
{for idx i in Array(3):
  g()
  (i, g() -> i)
}
```
and then get a runtime error about g being undefined, that is the desired behavior?

Whereas if they do the same thing with an array comprehension it will work?

**Comment by wcjohnson** [comment]
For the record, my own opinion is I'm fine with neither of those things working, or both of them working, but one working and another not for no apparent reason is something that bugs me a bit.

Mangling the AST is a foregone conclusion if we decide to support this pattern. We are already `insertBefore` or `insertAfter` a declaration or return in the implicit return case and the array case. This is essentially the exact same mangling.

**Comment by rattrayalex** [comment]
Hmm, this is tricky. Let me think about it. 
(you mean `insertAfter` I think btw)

**Comment by rattrayalex** [comment]
So, hoisted function declarations are a feature of JS that you probably have to know about to use. 

And people who use them are very likely to know that you can only hoist a top-level function (they may not think about it as a declaration/statement vs expression, but you know it can't be within something else). 

The form `a, b` is a little weird – it's like a faux-tuple – but it definitely looks like "a thing". And you can't declare a hoisted function "within a thing". 

I don't think people would be surprised to see this not work: 
```
  foo()
  (x, function foo() {...})
```
but I do think they would be surprised to see this not work:
```
  foo()
  function foo() {...}
```

the former says, "I'm sticking a function inside of this thing", the latter says, "I'm just declaring a function!"

LightScript generally seeks to sit *on top* of JavaScript – not to mess around with it. There are obviously times when that's not true, but it's kept to a minimum. 

Keeping Functions as declarations/expressions is following that spirit. 

Does that make sense?

**Comment by rattrayalex** [comment]
tl;dr, 

>  for no apparent reason

I think the reason is apparent

File: src/index.js [hunk 9]
```diff
@@ -883,34 +942,56 @@ export default function (babel) {
...
-          YieldExpression(yieldPath) {
-            throw yieldPath.buildCodeFrameError("`yield` is not allowed within Comprehensions.");
-          },
-          ReturnStatement(returnPath) {
-            throw returnPath.buildCodeFrameError("`return` is not allowed within Comprehensions.");
-          },
+        validateComprehensionLoopBody(path.get("loop.body"));
+
+        const id = path.scope.generateUidIdentifier("arr");
+        transformTails(path.get("loop"), true, function(expr) {
```
Comment (rattrayalex) [inline src/index.js]: would prefer an arrow function but nbd

**Comment by rattrayalex** [comment]
Other than the FunctionDeclaration/Expression bit, this looks good to go!

**Comment by wcjohnson** [comment]
Again for the record, I disagree philosophically, because we are letting people violate the JS pattern elsewhere. For instance, we know in regular JS:
```javascript
function f() {
  g()
  return function g() { }
}
```
doesn't work, because `function g()` is in expr context. Makes sense, fine and dandy.

But in LightScript, we are now allowing:
```javascript
f() ->
  g()
  g() -> {}
```
and it works fine. If we're thinking "light wrapper over JavaScript" most people would expect that to fail for the exact same reason, no? From the point of view of the actual user of the language, we are already turning expressions into declarations, whether or not the internals are actually doing that or not.

However, I think we are talking about the edgiest of edge cases here. We are basically talking about code that people should probably be shot for writing -- so I will revert that change.

**Comment by rattrayalex** [comment]
Cool, thanks. 

Yeah, this is sort of a problem with the whole idea of implicit returns... sometimes you want them, sometimes you don't. I'm really hoping to build tooling that encourages `: void` annotations whenever the retval isn't used. 

The problem with the implicit returns case is: 

```coffee
f() ->
  g()
  g() -> {}
```
the user might not have meant to return `g`; they might not be thinking of it as 
```js
function f() {
  g()
  return function g() { }
}
```
at all, but instead just as
```js
function f() {
  g()
  function g() { }
}
```

Of course, that breaks down when it comes to comprehensions, where the ending-thing is clearly going to be "returned". 

Anyway, you're definitely right that this isn't an edge-case worth spending much time on; which is largely why I'd really like the simplest/smallest/least intrusive solution. We need the "if its a declaration, return its name afterwards" code for implicit returns anyway, so reusing it when it's actually needed in comprehensions feels cleanest. 

Thanks for taking this on 😄 and excited to ship object comprehensions!

File: src/index.js [hunk 10]
```diff
@@ -183,22 +183,70 @@ export default function (babel) {
...
+          tailPath.replaceWith(getNewNode(tailPath.node.expression, tailPath));
+        }
+      } else if (tailPath.isVariableDeclaration()) {
+        // TODO: handle declarations.length > 1
+        // TODO: add linting to discourage
+        tailPath.insertAfter(getNewNode(tailPath.node.declarations[0].id, tailPath));
+      } else if (isFunctionDeclaration(tailPath.node)) {
+        // Need to transform exprs to statements since this is block context.
+        const nextNode = getNewNode(tailPath.node.id, tailPath);
+        if (t.is("Expression", nextNode)) {
```
Comment (rattrayalex) [inline src/index.js]: I think `t.isExpression` could be used next time, FYI (unless I'm missing something?) but I have no problem using this for now

Comment (wcjohnson) [reply src/index.js]: @rattrayalex Not sure if you're missing anything, TBH. I was looking into the Babel code to introspect node types and I found `t.is`. I pulled on the thread and I thought if I called that method it would correctly understand the LightScript type aliases. (So I wouldn't need e.g.
```
    return node && (t.is("FunctionDeclaration", node) || node.type === "NamedArrowDeclaration");
```
to  find out if something is a function declaration.

However, it turns out it doesn't work, and I can't figure out why it doesn't work NamedArrowDeclaration definitely has FunctionDeclaration as an alias, and the code for `t.is` definitely checks the aliases. I was totally mystified.

I bet t.isExpression() would be fine here so I will amend.

Comment (rattrayalex) [reply src/index.js]: Hmm, wouldn't this be necessary for the VariableDeclaration codepath too?

I guess the test case for that would be 
```
[for idx i in Array(10): 
  const x = f(i);
]
```
which reminds me we should definitely have a test for 
```
[for idx i in Array(10): 
  x = f(i);
]
```
which I have a hunch may not do what we want it to. 

Though I'm not sure in either case what we really want here. 

With implicit returns, it can be kind of nice to annotate your return value with a name – or at least, it can be annoying when doing so results in implicit returns turning off or breaking. I guess the same logic applies for comprehensions, so I think we want the same behavior, but it's a little weird (and should almost definitely be a linting error). 

If you just wanna get this shipped for now though I'm happy to merge and add the above in a follow-on

Comment (rattrayalex) [reply src/index.js]: Also, apologies for not noticing this earlier 😕

Comment (wcjohnson) [reply src/index.js]: My guess is the assignment expression will work but the declaration won't, will test.

Comment (wcjohnson) [reply src/index.js]: They both worked without change:
```
[for idx i in Array(10):
  now x = f(i)
]
```
```javascript
(() => {
  const _arr = [];

  const _arr2 = Array(10);

  for (let i = 0, _len = _arr2.length; i < _len; i++) {
    x = f(i);

    _arr.push(x)
  }

  return _arr;
})();
```
and
```
[for idx i in Array(10):
  const x = f(i)
]
```
```javascript
(() => {
  const _arr = [];

  const _arr2 = Array(10);

  for (let i = 0, _len = _arr2.length; i < _len; i++) {
    const x = f(i);

    _arr.push(x)
  }

  return _arr;
})();
```

Comment (wcjohnson) [reply src/index.js]: Looks right to me

File: test/fixtures/comprehensions/assignment-expr/actual.js [hunk 11]
```diff
@@ -0,0 +1,3 @@
+[for idx i in Array(10):
+  now x = f(i)
```
Comment (rattrayalex) [inline test/fixtures/comprehensions/assignment-expr/actual.js:2]: Actually, with `now`, it's perfectly fine to leave it as `arr.push(x = f(i));` – the case I was worried about was the auto-const `x = f(i)`

Comment (rattrayalex) [reply test/fixtures/comprehensions/assignment-expr/actual.js:2]: I'm fine with this output too though if the code is simpler

Comment (wcjohnson) [reply test/fixtures/comprehensions/assignment-expr/actual.js:2]: I expect we can change the output of `now` to look like that very easily, but for some reason `addImplicitReturns` extracts the LHS of the assignment operator when returning it.

AFAIK it's perfectly fine to `return x=3` or some such, but if you had some reason for returning the LHS in `addImplicitReturns` it would be good to know why that's there, and/or why this should be the same and/or different.

Right now it's just using the same code path.

Comment (wcjohnson) [reply test/fixtures/comprehensions/assignment-expr/actual.js:2]: Auto const works fine afaict.

**Review by rattrayalex — APPROVED** [review]
Awesome! Looks great! Good by you if I merge?

**Comment by wcjohnson** [comment]
@rattrayalex Squashed, good to merge as f ar as I can see.

File: test/fixtures/comprehensions/assignment-expr/actual.js (hunk 11 above)
Comment (rattrayalex) [reply test/fixtures/comprehensions/assignment-expr/actual.js:2]: Cool. Yeah it's there in `addImplicitReturns` from before that was moved to an `exit()` visitor, when `x = 3` was going to turn into `const x = 3`. It actually shouldn't be needed for returns anymore, but now we're relying on it for this. Fine to leave as-is I think.
